test/TestApps/PushNotificationsTestApp/main.cpp

```diff
@@ -0,0 +1,330 @@
+﻿#include "pch.h"
+#include <testdef.h>
+#include <iostream>
+#include <sstream>
+#include <wil/win32_helpers.h>
+#include <winrt/Windows.ApplicationModel.Background.h> // we need this for BackgroundTask APIs
+
+using namespace winrt;
+using namespace winrt::Microsoft::Windows::AppLifecycle;
+using namespace winrt::Microsoft::Windows::PushNotifications;
+using namespace winrt::Windows::ApplicationModel::Activation;
+using namespace winrt::Windows::ApplicationModel::Background; // BackgroundTask APIs
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::Storage::Streams;
+
+winrt::guid remoteId1(L"a2e4a323-b518-4799-9e80-0b37aeb0d225"); // Generated from ms.portal.azure.com
+winrt::guid remoteId2(L"CA1A4AB2-AC1D-4EFC-A132-E5A191CA285A"); // Dummy guid from visual studio guid tool generator
+
+PushNotificationRegistrationToken g_appToken = nullptr;
+
+bool ChannelRequestUsingNullRemoteId()
+{
+    try
+    {
+        auto channelOperation = PushNotificationManager::CreateChannelAsync(winrt::guid()).get();
+    }
+    catch (...)
+    {
+        return to_hresult() == E_INVALIDARG;
+    }
+    return false;
+}
+
+HRESULT ChannelRequestHelper(IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& channelOperation)
+{
+    if (channelOperation.wait_for(std::chrono::seconds(300)) != AsyncStatus::Completed)
+    {
+        channelOperation.Cancel();
+        return HRESULT_FROM_WIN32(ERROR_TIMEOUT); // timed out or failed
+    }
+
+    auto result = channelOperation.GetResults();
+    auto status = result.Status();
+    if (status != PushNotificationChannelStatus::CompletedSuccess)
+    {
+        return result.ExtendedError(); // did not produce a channel
+    }
+
+    result.Channel().Close();
+    return S_OK;
+}
+
+bool ChannelRequestUsingRemoteId()
+{
+    auto channelOperation = PushNotificationManager::CreateChannelAsync(remoteId1);
+    auto channelOperationResult = ChannelRequestHelper(channelOperation);
+
+    return channelOperationResult == S_OK;
+}
+
+// Verify calling channel close will fail when called twice.
+bool MultipleChannelClose()
+{
+    auto channelOperation = PushNotificationManager::CreateChannelAsync(remoteId1);
+    if (channelOperation.wait_for(std::chrono::seconds(300)) != AsyncStatus::Completed)
```

Comment:
300s? That's 5 minutes. Pretty big timeout. How long do you expect this to take in the average and worst case scenarios?

Reply:
We agreed with our team that 5mins (worst case) will be used here. Raymond is also okay with it. So for now this is the best value we can use here.

Reply:
I'm not a big fan of a 5-minute time out. Nobody will be that patient. But I thought CreateChannelAsync already had a timeout. Why not just await the CreateChannelCall and use the built-in timeout? Isn't that what we should be recommending?

Reply:
@oldnewthing yes ideally here we should be recommending await on the CreateChannelAsync call and just get the result. As these are unittests above we sort of opted for wait_for here instead of await. I will evaluate this again and see if we can improvise this in the upcoming PRs. Thanks.

Reply:
Does this indicate a deficiency in our API if even our own tests cannot use the API in the normal way? Should CreateChannelAsync have a custom timeout parameter?